src/rust/src/x509.rs

```diff
@@ -391,6 +411,296 @@ fn load_der_x509_certificate(py: pyo3::Python<'_>, data: &[u8]) -> PyAsn1Result<
+        &self,
+        py: pyo3::Python<'p>,
+    ) -> Result<&'p pyo3::PyAny, PyAsn1Error> {
+        let sig_oids_to_hash = py
+            .import("cryptography.hazmat._oid")?
+            .getattr("_SIG_OIDS_TO_HASH")?;
+        let hash_alg = sig_oids_to_hash.get_item(self.signature_algorithm_oid(py)?);
+        match hash_alg {
+            Ok(data) => Ok(data),
+            Err(_) => Err(PyAsn1Error::from(pyo3::PyErr::from_instance(
+                py.import("cryptography.exceptions")?.call_method1(
+                    "UnsupportedAlgorithm",
+                    (format!(
+                        "Signature algorithm OID: {} not recognized",
+                        self.raw.borrow_value().signature_alg.oid.to_string()
+                    ),),
+                )?,
+            ))),
+        }
+    }
+
+    #[getter]
+    fn signature_algorithm_oid<'p>(&self, py: pyo3::Python<'p>) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        py.import("cryptography.x509")?.call_method1(
+            "ObjectIdentifier",
+            (self.raw.borrow_value().signature_alg.oid.to_string(),),
+        )
+    }
+
+    fn public_bytes<'p>(
+        &self,
+        py: pyo3::Python<'p>,
+        encoding: &pyo3::PyAny,
+    ) -> pyo3::PyResult<&'p pyo3::types::PyBytes> {
+        let encoding_class = py
+            .import("cryptography.hazmat.primitives.serialization")?
+            .getattr("Encoding")?;
+
+        let result = asn1::write_single(self.raw.borrow_value());
+        if encoding == encoding_class.getattr("DER")? {
+            Ok(pyo3::types::PyBytes::new(py, &result))
+        } else if encoding == encoding_class.getattr("PEM")? {
+            let pem = pem::encode_config(
+                &pem::Pem {
+                    tag: "CERTIFICATE REQUEST".to_string(),
+                    contents: result,
+                },
+                pem::EncodeConfig {
+                    line_ending: pem::LineEnding::LF,
+                },
+            )
+            .into_bytes();
+            Ok(pyo3::types::PyBytes::new(py, &pem))
+        } else {
+            Err(pyo3::exceptions::PyTypeError::new_err(
+                "encoding must be Encoding.DER or Encoding.PEM",
+            ))
+        }
+    }
+
+    fn get_attribute_for_oid<'p>(
+        &self,
+        py: pyo3::Python<'p>,
+        oid: &pyo3::PyAny,
+    ) -> pyo3::PyResult<&'p pyo3::PyAny> {
+        // TODO: Convert ObjectIdentifier in python to a rust class so this can be more efficient
+        let oid_str = oid
+            .getattr("dotted_string")?
+            .downcast::<pyo3::types::PyString>()?
+            .to_string();
+        let rust_oid = asn1::ObjectIdentifier::from_string(&oid_str).unwrap();
+        for attribute in self.raw.borrow_value().csr_info.attributes.clone() {
+            if rust_oid == attribute.type_id {
+                check_attribute_length(attribute.values.clone())?;
+                if let Some(val) = attribute.values.clone().next() {
```

Comment:
This can just be `val = attributes.values.clone().next().unwrap()` -- there's always an element in the iterator due to teh `check_attribute_length`